Add plugin setting options for selecting Gerrit review label

juhasipo opened the pull request:
Previously the only way for this plugin to mark a Gerrit build as passed has been the `Code-Review` field. This field is usually used for the actual code review made by a person while the build automation has used the `Verified` field. This pull request enables the admin to choose the used field via plugin settings.

ketan commented on src/main/java/com/tw/go/plugin/BuildStatusNotifierPlugin.java:
```diff
@@ -102,6 +106,9 @@ private GoPluginApiResponse handleGetPluginSettingsConfiguration() {
         response.put(PLUGIN_SETTINGS_USERNAME, createField("Username", null, true, false, "2"));
         response.put(PLUGIN_SETTINGS_PASSWORD, createField("Password", null, true, true, "3"));
         response.put(PLUGIN_SETTINGS_OAUTH_TOKEN, createField("OAuth Token", null, true, true, "4"));
+        if (provider.pluginId().equals(GerritProvider.PLUGIN_ID)) {
```
The `GerritProvider` class won't be available inside other plugins, causing a CNFE here. Do you feel it's better to push this behaviour into the `Provider` interface, the provider could be passed the response object so it can add more fields into it, and determine the correct index it needs to start from, based on the size of the existing fields.

juhasipo replied:
You're right, it would be much cleaner if the providers would contain the configurations especially if the template name is moved to there too. And if in the future there are more general or plugin specific configurations added that method would become really messy. I might as well make the improvement already at this point.

ketan commented on src/main/java/com/tw/go/plugin/BuildStatusNotifierPlugin.java:
```diff
@@ -55,6 +58,7 @@ public BuildStatusNotifierPlugin() {
             Class<?> providerClass = Class.forName(properties.getProperty("provider"));
             Constructor<?> constructor = providerClass.getConstructor();
             provider = (Provider) constructor.newInstance();
+            templateName = properties.getProperty("settings.template");
```
This IMO should come from the provider, to keep use of configuration files to a minimum.

juhasipo replied:
That actually would be a better way to do this. I'll move that to the provider.

ketan commented on src/main/java/com/tw/go/plugin/provider/Provider.java:
```diff
@@ -2,11 +2,16 @@
     public void updateStatus(String url, PluginSettings pluginSettings, String branch, String revision, String pipelineStage,
                              String result, String trackbackURL) throws Exception;
+
+    public List<Map<String, Object>> validateConfig(Map<String, Object> fields);
```
👍

ketan commented:
@juhasipo I'm not too familiar with the gerrit API, but your PR looks good in terms of what it does. I feel there's some scope of improvement in terms of not changing the shared code aka `BuildStatusNotifierPlugin` to avoid the next person from adding something similar for other git providers.

If there's specific behavior or additions to the request/response that needs to be different per git provider, it's perhaps better to push that into the `Provider` interface, even if it may mean no-op implementations for other providers that don't need those capabilities.

What do you think?

juhasipo commented:
@ketan 
Thank you for your feedback. I think that you have a valid point about not changing the shared code and moving provider specific code to the provider implementations. I'll fix my code so that it won't affect the `BuildStatusNotifierPlugin` code.

Should creating the plugin settings be also moved the providers too? Now the `PluginSettings` object is created in the `BuildStatusNotifierPlugin` and it contains some Gerrit provider specific code. The only place the `PluginSettings` is actually used is in the providers. The `BuildStatusNotifierPlugin` creates the settings object now because the `GoApplicationAccessor` is accessible there.

In the future the `PluginSettings` class may grow even more and contain more plugin specific settings that are only used in one plugin. Maybe all the plugins should have their own settings classes so that the `PluginSettings` wouldn't have any unnecessary properties.

What is your opinion on this?

ketan commented:
> Should creating the plugin settings be also moved the providers too? Now the PluginSettings object is created in the BuildStatusNotifierPlugin and it contains some Gerrit provider specific code. The only place the PluginSettings is actually used is in the providers. The BuildStatusNotifierPlugin creates the settings object now because the GoApplicationAccessor is accessible there.
> 
> In the future the PluginSettings class may grow even more and contain more plugin specific settings that are only used in one plugin. Maybe all the plugins should have their own settings classes so that the PluginSettings wouldn't have any unnecessary properties.

I think there's a case to be made for having a `PluginSettings` class that may be overridden for specific settings that each plugin may require. Not all plugins need it, so far :)

Something along these lines (example code I just typed, may not compile) —

```java
public interface Provider {
  // add a method for the provider to provide its own settings object
  PluginSettings getPluginSettings(String responseBody);
  ...
}

public class DefaultProvider implements Provider {
  public PluginSettings getPluginSettings(String responseBody) {
    // See https://github.com/gocd-contrib/gocd-build-status-notifier/blob/4ff7ebae98c58fed21973d9bc91d5594ecbdd854/src/main/java/com/tw/go/plugin/BuildStatusNotifierPlugin.java#L133-L135
    return new PluginSettings(....);
  }
  ...
}

public class FooPluginSettings extends PluginSettings {
  // implement any additional settings here
}

public class FooGitProvider extends DefaultProvider {
  public PluginSettings getPluginSettings(){
    // note the specific subtype being returned
    return new FooPluginSettings(...);
  }

  public void updateStatus(PluginSettings s, ...) {
    // The provider may cast it to its own settings object
    // Using generics could probably improve this a lot, but you get the idea :)
    FooPluginSettings pluginSettings = (FooPluginSettings) s;
  }

  ...
}
```

Do you feel this is reasonable, is there a better way?

juhasipo commented:
@ketan 
The other night when I was looking at the code I missed the `String serverBaseURLToUse = pluginSettings.getServerBaseURL();` line in the `BuildStatusNotifierPlugin#handleStageNotification` method. Originally I thought that the whole plugin setting handling could've been done inside the providers without exposing it to outside of the class. But since the `PluginSettings` object is used in the `BuildStatusNotifierPlugin` my original idea doesn't work.

I implemented it now the way you described. In addition to that I moved the view configurations to separate `PluginConfigurationView` classes so that the provider classes don't start growing too much. Also the default configuration view can be now reused for the GitHub and Stash providers.

ketan commented:
Hi,

I'm reviewing this now. I'm sorry it took me so long.
